synapse/third_party_rules/access_rules.py

```diff
@@ -442,14 +470,28 @@ def _on_membership_or_invite_restricted(self, event: EventBase) -> bool:
         invitee_domain = get_domain_from_id(event.state_key)
         return invitee_domain not in self.domains_forbidden_when_restricted
 
-    def _on_membership_or_invite_unrestricted(self) -> bool:
+    def _on_membership_or_invite_unrestricted(
+        self, event: EventBase, state_events: StateMap[EventBase]
+    ) -> bool:
         """Implements the checks and behaviour specified for the "unrestricted" rule.
 
         "unrestricted" currently means that every event is allowed.
 
         Returns:
             True if the event can be allowed, False otherwise.
         """
+        # If this is a join or invite for a room that's not listed in the public room list,
+        # then deny if the target is an external user
+        if event.type == EventTypes.Member and event.membership == Membership.JOIN:
+            target_user = UserID.from_string(event.state_key)
+            if (
+                target_user.domain in self.domains_forbidden_when_restricted
+                and not self.module_api.public_room_list_manager.room_is_in_public_room_list(
+                    event.room_id
+                )
+            ):
```

Comment:
I don't think this is the right check here.

We have to deny the join for the room if the target is an external and if they are not invited (even if the join_rule is public)

if (target_domain in self.domains_forbidden_when_restricted
   and event.prev_content. membership != Membership.INVITE) :

Reply:
This should be fixed in https://github.com/matrix-org/synapse-dinsic/pull/63/commits/7a697a6b2925b396b311985d04d1a8ae5545688d :)